sql: throw pgerror instead of assertion error for bad OID strings

otan opened the pull request:
Resolves #69907

Release note (sql change): Return a pgerror with code 42704 when an
invalid cast to OID is made. This previously threw an assertion error.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/70454)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
tiny nits!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70454#-:-Mk5J16gBCddoIkRVBZg:b3kx3f0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[pkg/sql/resolve_oid.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70454#-Mk5IY6B01Oosy8ysT6d:-Mk5IY6EDI6ju3cAGcTB:b-zd67jh) ([raw file](https://github.com/cockroachdb/cockroach/blob/360efc9ec0e4bb4a8691aa66b895610c34994a05/pkg/sql/resolve_oid.go#L59)):*
> ```Go
> 	if !ok {
> 		return nil, pgerror.Newf(
> 			pgcode.UndefinedObject,
> ```

nit: looks like PG uses InvalidTextRepresentation (22P02) for this
___
*[pkg/sql/resolve_oid.go, line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70454#-Mk5IcB59Rc4LEbxz_ut:-Mk5IcB59Rc4LEbxz_uu:b-info9u) ([raw file](https://github.com/cockroachdb/cockroach/blob/360efc9ec0e4bb4a8691aa66b895610c34994a05/pkg/sql/resolve_oid.go#L60)):*
> ```Go
> 		return nil, pgerror.Newf(
> 			pgcode.UndefinedObject,
> 			"invalid input syntax for type %s: %s",
> ```

should the second one be a `%q`?
___
*[pkg/sql/resolve_oid.go, line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70454#-Mk5Iufj15Grih78rw1P:-Mk5Iufj15Grih78rw1Q:b-fr5k7y) ([raw file](https://github.com/cockroachdb/cockroach/blob/360efc9ec0e4bb4a8691aa66b895610c34994a05/pkg/sql/resolve_oid.go#L62)):*
> ```Go
> 			"invalid input syntax for type %s: %s",
> 			resultType,
> 			toResolve.String(),
> ```

should this use `tree.AsStringWithFlags(toResolve, tree.FmtBareStrings)` (or is is `FmtBareIdentifiers`?)


<!-- Sent from Reviewable.io -->

otan committed:
sql: throw pgerror instead of assertion error for bad OID strings

Release note (sql change): Return a pgerror with code 22P02 when an
invalid cast to OID is made. This previously threw an assertion error.

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70454)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[pkg/sql/resolve_oid.go, line 59 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70454#-Mk5IY6B01Oosy8ysT6d:-Mk5Jehd-t3jFp5K8_sg:b-zc5u0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/360efc9ec0e4bb4a8691aa66b895610c34994a05/pkg/sql/resolve_oid.go#L59)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: looks like PG uses InvalidTextRepresentation (22P02) for this
</blockquote></details>

huh, i thought it was one but i could've been wrong
___
*[pkg/sql/resolve_oid.go, line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70454#-Mk5IcB59Rc4LEbxz_ut:-Mk5Je8B4Q6otXQRd_O8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/360efc9ec0e4bb4a8691aa66b895610c34994a05/pkg/sql/resolve_oid.go#L60)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should the second one be a `%q`?
</blockquote></details>

Done.
___
*[pkg/sql/resolve_oid.go, line 62 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70454#-Mk5Iufj15Grih78rw1P:-Mk5JeXA3zeVQ2BdwUHv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/360efc9ec0e4bb4a8691aa66b895610c34994a05/pkg/sql/resolve_oid.go#L62)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should this use `tree.AsStringWithFlags(toResolve, tree.FmtBareStrings)` (or is is `FmtBareIdentifiers`?)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
ty for fixing!!

> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70454#-:-Mk5VDN-ASAqnOTR292F:b-yqrbae)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)



<!-- Sent from Reviewable.io -->

otan commented:
bors r=rafiss

rafiss commented:
blathers backport 21.2 21.1
